File: scala/src/test/scala/bank/solution/WithdrawExampleTests.scala
package bank.solution

import bank.solution.WithdrawExampleTests._
import bank.{Account, AccountService, Amount, Withdraw}
import org.scalatest.EitherValues
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.must.Matchers.{be, contain, have, startWith}
import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
import org.scalatest.prop.TableDrivenPropertyChecks
import org.scalatest.prop.Tables.Table
import org.scalatestplus.scalacheck.Checkers

import java.time.LocalDate
import java.util.UUID

class WithdrawExampleTests
    extends AnyFlatSpec
    with TableDrivenPropertyChecks
    with EitherValues {

  "withdraw" should "pass examples" in {
    forAll(`withdraw examples`) {
      (
          balance,
          isOverdraftAuthorized,
          maxWithdrawal,
          withdrawAmount,
          expectedResult
      ) =>
        val command = toWithdraw(withdrawAmount)
        val debitedAccount =
          AccountService.withdraw(
            Account(balance, isOverdraftAuthorized, maxWithdrawal),
            command
          )

        expectedResult match {
          case Left(expectedErrorMessage) =>
            debitedAccount.left.value should startWith(expectedErrorMessage)
          case Right(expectedBalance) =>
            debitedAccount.value.balance should be(expectedBalance)
            debitedAccount.value.withdraws should contain(command)
            debitedAccount.value.withdraws should have length 1
        }
    }
  }
}

object WithdrawExampleTests {
  private val `withdraw examples` =
    Table(
      (
        "balance",
        "isOverdraftAuthorized",
        "maxWithdrawal",
        "withdrawAmount",
        "expectedResult"
      ),
      (10_000, false, 1200, 1199.99, Right(8800.01)),
      (0, true, 500, 50d, Right(-50)),
      (1, false, 1200, 1199.99, Left("Insufficient balance to withdraw")),
      (0, false, 500, 50d, Left("Insufficient balance to withdraw")),
      (10_000, true, 1200, 1200.0001, Left("Amount exceeding your limit of")),
      (0, false, 500, 500d, Left("Amount exceeding your limit of"))
    )

  private def toWithdraw(amount: Double): Withdraw = {
    Withdraw(
      UUID.randomUUID(),
      Amount.from(amount).get,
      LocalDate.now()
    )
  }
}

File: java/src/test/java/bank/solution/WithdrawExampleTests.java
package bank.solution;

import bank.Account;
import bank.AccountService;
import bank.Amount;
import bank.Withdraw;
import io.vavr.collection.List;
import io.vavr.control.Either;
import lombok.val;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.time.LocalDate;
import java.util.UUID;
import java.util.stream.Stream;

import static io.vavr.control.Either.left;
import static io.vavr.control.Either.right;
import static org.assertj.core.api.Assertions.assertThat;

class WithdrawExampleTests {
    public static Stream<Arguments> examples() {
        return Stream.of(
                Arguments.of(10_000, false, 1200, 1199.99, right(8800.01)),
                Arguments.of(0, true, 500, 50, right(-50d)),
                Arguments.of(1, false, 1200, 1199.99, left("Insufficient balance to withdraw")),
                Arguments.of(0, false, 500, 50, left("Insufficient balance to withdraw")),
                Arguments.of(10_000, true, 1200, 1200.0001, left("Amount exceeding your limit of")),
                Arguments.of(0, false, 500, 500, left("Amount exceeding your limit of"))
        );
    }

    @ParameterizedTest
    @MethodSource("examples")
    void withdraw_should(
            double balance,
            boolean isOverdraftAuthorized,
            double maxWithdrawal,
            double withdrawAmount,
            Either<String, Double> expectedResult
    ) {
        val command = new Withdraw(UUID.randomUUID(), Amount.from(withdrawAmount).get(), LocalDate.now());
        val account = new Account(balance, isOverdraftAuthorized, maxWithdrawal, List.empty());
        val result = AccountService.withdraw(account, command);

        expectedResult
                .peek(expectedBalance -> {
                    val debitedAccount = result.get();

                    assertThat(debitedAccount.balance()).isEqualTo(expectedBalance);
                    assertThat(debitedAccount.withdraws())
                            .contains(command)
                            .hasSize(1);
                })
                .peekLeft(expectedErrorMessage -> assertThat(result.getLeft()).startsWith(expectedErrorMessage));
    }
}
